src/test/scala/org/broadinstitute/dsde/workbench/leonardo/service/AuthProviderSpec.scala

```diff
@@ -0,0 +1,177 @@
+
+      //can't make a cluster
+      val clusterCreateException = leo.createCluster(userInfo, project, name1, testClusterRequest).failed.futureValue
+      clusterCreateException shouldBe a [AuthorizationError]
+      clusterCreateException.asInstanceOf[AuthorizationError].statusCode shouldBe StatusCodes.Unauthorized
+
+      //can't get details on an existing cluster
+      //poke a cluster into the database so we actually have something to look for
+      dbFutureValue { _.clusterQuery.save(c1, gcsPath("gs://bucket1"), None) }
+
+      val clusterGetResponseException = leo.getActiveClusterDetails(userInfo, c1.googleProject, c1.clusterName).failed.futureValue
+      clusterGetResponseException shouldBe a [ClusterNotFoundException]
+      clusterGetResponseException.asInstanceOf[ClusterNotFoundException].statusCode shouldBe StatusCodes.NotFound
+
+      //list all clusters should be fine, but empty
+      val emptyList = leo.listClusters(userInfo, Map()).futureValue
+      emptyList shouldBe 'empty
+
+      //connect to cluster
+      val httpRequest = HttpRequest(GET, Uri(s"/notebooks/$googleProject/$clusterName"))
+      val clusterNotFoundException = proxy.proxy(userInfo, GoogleProject(googleProject), ClusterName(clusterName), httpRequest, tokenCookie).failed.futureValue
+      clusterNotFoundException shouldBe a [ClusterNotFoundException]
+
+      //destroy cluster
+      val clusterNotFoundAgain = leo.deleteCluster(userInfo, project, name1).failed.futureValue
+      clusterNotFoundAgain shouldBe a [ClusterNotFoundException]
+    }
+
+    "should give you a 401 if you can see a cluster's details but can't do the more specific action" in isolatedDbTest {
+      val readOnlyProvider = new MockLeoAuthProvider(config.getConfig("auth.readOnlyProviderConfig"))
+      val leo = leoWithAuthProvider(readOnlyProvider)
+      val proxy = proxyWithAuthProvider(readOnlyProvider)
+
+      //poke a cluster into the database so we actually have something to look for
+      dbFutureValue { _.clusterQuery.save(c1, gcsPath("gs://bucket1"), None) }
+
+      // status should work for this user
+      val clusterStatus = leo.getActiveClusterDetails(userInfo, project, name1).futureValue
+      clusterStatus shouldBe c1
+
+      // list should work for this user
+      //list all clusters should be fine, but empty
+      val clusterList = leo.listClusters(userInfo, Map()).futureValue
+      clusterList shouldBe Seq(c1)
+
+      //connect should 401
+      val httpRequest = HttpRequest(GET, Uri(s"/notebooks/$googleProject/$clusterName"))
+      val clusterAuthException = proxy.proxy(userInfo, GoogleProject(googleProject), ClusterName(clusterName), httpRequest, tokenCookie).failed.futureValue
+      clusterAuthException shouldBe a [AuthorizationError]
+
+      //destroy should 401 too
+      val clusterDestroyException = leo.deleteCluster(userInfo, project, name1).failed.futureValue
+      clusterDestroyException shouldBe a [AuthorizationError]
+    }
+  }
```

Comment:
Can we add test cases for create/delete notifications?

Reply:
done